README.md
# libp2p-rpc

An RPC module for Libp2p.

## Table of Contents

- [Install](#install)
- [Usage](#usage)
- [API](#api)
  - [createRPC](#createrpc)
  - [RPC](#rpc)
    - [start](#start)
    - [stop](#stop)
    - [addMethod](#addmethod)
    - [hasMethod](#hasmethod)
    - [removeMethod](#removemethod)
    - [request](#request)
    - [notify](#notify)
  - [RPCException](#rpcexception)
- [Logging](#logging)
- [Tests](#tests)
- [Protocol Design](#protocol-design)
- [To-Do](#to-do)

## Install

```
npm i @organicdesign/libp2p-rpc
```

## Usage

```javascript
import { createRPC, RPCException } from "@organicdesign/libp2p-rpc";

const rpc = createRPC([options])(libp2p);

// Add an RPC method.
rpc.addMethod("log", (params, peerId) => {
	console.log(`received ${params} from ${peerId.toString()}`);

	const somethingHasGoneWrong = false;

	if (somethingHasGoneWrong) {
		// Respond with an error with a message and code.
		throw new RPCException("Something went wrong", 0);
	}

	// Return a response
	return new Uint8Array([3]);
});

// Start the rpc module.
await rpc.start();

// Call the 'log' method on the connected peer with ID peerId with parameter.
const response = await rpc.request(peerId, "log", new Uint8Array([1]));

console.log("Response: ", response); // Response:  Uint8Array [ 3 ]

// Call the 'log' method without expecting a response.
rpc.notify(peerId, "log", new Uint8Array([2]));

// Stop the rpc module.
await rpc.stop();
```

## API

### createRPC

```javascript
createRPC([options])(libp2p);
```

- `options` `<Object>` An optional object with the following properties:
  - `protocol` `<string>` A string which specifies the name of the protocol. Default: `"/libp2p-rpc/0.0.1"`.
  - `timeout` `<number>` The timeout in milliseconds for requests. Set to `-1` to disable. Default: `5000` (5s)
- `libp2p` `<Libp2p>` The libp2p instance.
- Returns: `<RPC>` The RPC instance.

Creates a Libp2p RPC module.

### RPC

```javascript
createRPC(libp2p, [options]);
```

- `options` `<Object>` An optional object with the following properties:
  - `protocol` `<string>` A string which specifies the name of the protocol. Default: `"/libp2p-rpc/0.0.1"`.
  - `timeout` `<number>` The timeout in milliseconds for requests. Set to `-1` to disable. Default: `5000` (5s)
- `libp2p` `<Libp2p>` The libp2p instance.

The RPC class. It is not recommended to instanciate it directly but rather use the `createRPC` function.

#### start

```javascript
rpc.start();
```

- Returns: `<Promise>`

Start the RPC module, resolves when it has finished starting.

#### stop

```javascript
rpc.stop();
```

- Returns: `<Promise>`

Stop the RPC module, resolves when it has finished stopping. This will clear all remote procedure call methods.

#### addMethod

```javascript
rpc.addMethod(name, method);
```

- `name` `<string>` The name of the method to add.
- `method` `<RPCMethod>` The RPC method to add. RPCMethod is a method with the parameters:
  - `params` `<Uint8Array> | <undefined>` The paramters this method was called with.
  - `sender` `<PeerId>` The peer that called this method.

Add a remote procedure call method. Methods can be added dynamically (before or after the `start` call).

#### hasMethod

```javascript
rpc.hasMethod(name);
```

- `name` `<string>` The name of the method to check.
- Returns: `<boolean>` True if there is a method handled by that name.

Check if a remote procedure call method exists.

#### removeMethod

```javascript
rpc.removeMethod(name);
```

- `name` `<string>` The name of the method to remove.
- Returns: `<boolean>` True if the method did exist and was removed.

Remove a remote procedure call method. Methods can be removed dynamically (before or after the `start` call).

#### request

```javascript
rpc.request(peerId, name, [params]);
```

- `peerId` `<PeerId>` The ID of the peer to request the call on.
- `name` `<string>` The name of the method to execute.
- `params` `<Uint8Array>` The paramters to call the method with.
- Returns: `<Promise>`

Request a remote peer to execute a method. The promise resolves with the result of the RPC method or rejects with an `RPCException` if the method fails or it fails to send the request (the peer is not connected) or it times out.

#### notify

```javascript
rpc.notify(peerId, name, [params]);
```

- `peerId` `<PeerId>` The ID of the peer to request the call on.
- `name` `<string>` The name of the method to execute.
- `params` `<Uint8Array>` The paramters to call the method with.

Request a remote peer to execute a method. This executes synchronously, does not return any value and will not throw even if it fails to send the request.

### RPCException

```javascript
new RPCException(message, code, [data]);
```

- `message` `<string>` The error message.
- `code` `<number>` The error code.
- `data` `<Uint8Array>` Optional error data.

The `RPCException` class is an extention of the native `Error` class with additional properties to hold an error code and error data. You can throw `RPCException`s inside RPC methods to return additional data back to the remote caller.

The pre-defined error codes are:

- `-32601` `Method not found` Thrown when a method is called on a node that does not have a matching handler.
- `-32000` Thrown when the request could not be sent to the remote peer. The message is not static and is the reason why the request message could not be sent. This can be caused by not being connected to it.
- `-32001` `RPC module stopped` Thrown when a request is aborted due to the module stopping.
- `-32002` Thrown when the method throws an error but does not define a code. The message is not static and is the message of the error that was thrown.
- `-32003` `Request timed out` Thrown when a request timed out.

## Logging

The logger has the following namespaces:

* `libp2p:rpc` - Logs general actions like starting, stopping and incoming/outgoing rpc calls.

To enable logging in nodejs add the following environment variable (by prefixing the start command):

```
DEBUG=libp2p:rpc*
```

Or in the browser:

```javascript
localStorage.setItem("debug", "libp2p:rpc*");
```

## Protocol Design

This module is loosely based on the [JSON-RPC 2.0 Specification](https://www.jsonrpc.org/specification) but altered to work efficiently over libp2p.

The parameters are bytes instead of an object or array, the response is also bytes and all the fields are encoded into binary using protocol buffers instead of JSON.

## Tests

To run the test suite:

```
npm run test
```

## To-Do

- [ ] More tests.


CHANGELOG.md
## [0.2.2](https://github.com/organicdesign/libp2p-rpc/compare/v0.2.1...v0.2.2) (2023-02-17)

### Fixed

* Fixed security vulnerabilities in packages. (undici)

## [0.2.1](https://github.com/organicdesign/libp2p-rpc/compare/v0.2.0...v0.2.1) (2023-02-02)

### Fixed

* Fixed security vulnerabilities in packages.

## [0.2.0](https://github.com/organicdesign/libp2p-rpc/compare/v0.1.0...v0.2.0) (2023-01-24)

### Added

* Add additional methods for RPC handlers:
  * `hasMethod`
  * `removeMethod`

## 0.1.0 (2023-01-20)

### Added

* Functionality to execute remote procedures over libp2p.


test/RPC.spec.ts
import { createRSAPeerId } from "@libp2p/peer-id-factory";
import { mockRegistrar, mockConnectionManager, mockNetwork } from "@libp2p/interface-mocks";
import { start } from "@libp2p/interfaces/startable";
import { stubInterface } from "ts-sinon";
import type { PeerId } from "@libp2p/interface-peer-id";
import type { ConnectionManager } from "@libp2p/interface-connection-manager";
import type { Libp2p } from "@libp2p/interface-libp2p";
import { RPCComponents, RPC, createRPC } from "../src/index.js";

interface TestRPCComponents extends RPCComponents {
	peerId: PeerId
	dial: Libp2p["dial"]
}

const createComponents = async (): Promise<TestRPCComponents> => {
	const oldComponents = {
		peerId: await createRSAPeerId({ bits: 512 }),
		registrar: mockRegistrar(),
		connectionManager: stubInterface<ConnectionManager>() as ConnectionManager
	};

	oldComponents.connectionManager = mockConnectionManager(oldComponents);

	const components: TestRPCComponents = {
		peerId: oldComponents.peerId,
		dial: (peerId) => oldComponents.connectionManager.openConnection(peerId),
		handle: (protocol: string, handler) => oldComponents.registrar.handle(protocol, handler),
		unhandle: (protocol: string) => oldComponents.registrar.unhandle(protocol),
		getConnections: () => oldComponents.connectionManager.getConnections()
	};

	await start(...Object.entries(components));

	mockNetwork.addNode(oldComponents);

	return components;
};

let localComponents: TestRPCComponents;
let localRpc: RPC;
let remoteComponents: TestRPCComponents;
let remoteRpc: RPC;

beforeEach(async () => {
	localComponents = await createComponents();
	localRpc = createRPC()(localComponents);
	remoteComponents = await createComponents();
	remoteRpc = createRPC()(remoteComponents);
});

afterEach(async () => {
	await localRpc.stop();
	await remoteRpc.stop();
});

describe("startable interface", () => {
	it("is not started after creation", async () => {
		expect(localRpc.isStarted()).toBe(false);
	});

	it("starts", async () => {
		await localRpc.start();

		expect(localRpc.isStarted()).toBe(true);
	});

	it("stops", async () => {
		await localRpc.start();
		await localRpc.stop();

		expect(localRpc.isStarted()).toBe(false);
	});
});

describe("methods", () => {
	it("adds methods", () => {
		const method = "test";

		// Check something hasn't persisted at the start.
		expect(localRpc.hasMethod(method)).toBe(false);

		localRpc.addMethod(method, () => {});

		expect(localRpc.hasMethod(method)).toBe(true);
	});

	it("removes methods", () => {
		const method = "test";

		// Check something hasn't persisted at the start.
		expect(localRpc.hasMethod(method)).toBe(false);

		localRpc.addMethod(method, () => {});
		localRpc.removeMethod(method);

		expect(localRpc.hasMethod(method)).toBe(false);
	});
});

describe("rpc", () => {
	beforeEach(async () => {
		await localRpc.start();
		await remoteRpc.start();

		await remoteComponents.dial(localComponents.peerId);
	});

	afterEach(async () => {
		await Promise.all(localComponents.getConnections().map(c => c.close()));
		await Promise.all(remoteComponents.getConnections().map(c => c.close()));
	});

	it("calls the handler method on notifications", async () => {
		const params = new Uint8Array([1, 2, 3]);

		const dataPromise: Promise<{ params: Uint8Array, sender: PeerId }> = new Promise(resolve => {
			localRpc.addMethod("test", (params: Uint8Array, sender: PeerId) => {
				resolve({ params, sender });
			});
		});

		remoteRpc.notify(localComponents.peerId, "test", params);

		const data = await dataPromise;

		expect(data.sender.equals(remoteComponents.peerId)).toBe(true);
		expect(data.params).toStrictEqual(params);
	});

	it("calls the handler method on requests and returns the result", async () => {
		const params = new Uint8Array([1, 2, 3]);
		const response = new Uint8Array([2, 3, 4]);

		const dataPromise: Promise<{ params: Uint8Array, sender: PeerId }> = new Promise(resolve => {
			localRpc.addMethod("test", (params: Uint8Array, sender: PeerId) => {
				resolve({ params, sender });

				return response;
			});
		});

		const methodResponse = await remoteRpc.request(localComponents.peerId, "test", params);

		const data = await dataPromise;

		expect(data.sender.equals(remoteComponents.peerId)).toBe(true);
		expect(data.params).toStrictEqual(params);
		expect(methodResponse).toStrictEqual(response);
	});
});


test/index.spec.ts
import { createRSAPeerId } from "@libp2p/peer-id-factory";
import { mockRegistrar, mockConnectionManager, mockNetwork } from "@libp2p/interface-mocks";
import { start } from "@libp2p/interfaces/startable";
import { stubInterface } from "ts-sinon";
import type { PeerId } from "@libp2p/interface-peer-id";
import type { ConnectionManager } from "@libp2p/interface-connection-manager";
import { RPCComponents, RPC, createRPC } from "../src/index.js";

const createComponents = async (): Promise<RPCComponents & { peerId: PeerId }> => {
	const components: RPCComponents & { peerId: PeerId } = {
		peerId: await createRSAPeerId({ bits: 512 }),
		registrar: mockRegistrar(),
		connectionManager: stubInterface<ConnectionManager>()
	};

	components.connectionManager = mockConnectionManager(components);

	await start(...Object.entries(components));

	mockNetwork.addNode(components);

	return components;
};

let localComponents: RPCComponents & { peerId: PeerId };
let localRpc: RPC;
let remoteComponents: RPCComponents & { peerId: PeerId };
let remoteRpc: RPC;

beforeEach(async () => {
	localComponents = await createComponents();
	localRpc = createRPC()(localComponents);
	remoteComponents = await createComponents();
	remoteRpc = createRPC()(remoteComponents);
});

afterEach(async () => {
	await localRpc.stop();
	await remoteRpc.stop();
});

describe("startable interface", () => {
	it("is not started after creation", async () => {
		expect(localRpc.isStarted()).toBe(false);
	});

	it("starts", async () => {
		await localRpc.start();

		expect(localRpc.isStarted()).toBe(true);
	});

	it("stops", async () => {
		await localRpc.start();
		await localRpc.stop();

		expect(localRpc.isStarted()).toBe(false);
	});
});

describe("rpc", () => {
	beforeEach(async () => {
		await localRpc.start();
		await remoteRpc.start();

		await remoteComponents.connectionManager.openConnection(localComponents.peerId);
	});

	afterEach(async () => {
		await Promise.all(localComponents.connectionManager.getConnections().map(c => c.close()));
		await Promise.all(remoteComponents.connectionManager.getConnections().map(c => c.close()));
	});

	it("calls the handler method on notifications", async () => {
		const params = new Uint8Array([1, 2, 3]);

		const dataPromise: Promise<{ params: Uint8Array, sender: PeerId }> = new Promise(resolve => {
			localRpc.addMethod("test", (params: Uint8Array, sender: PeerId) => {
				resolve({ params, sender });
			});
		});

		remoteRpc.notify(localComponents.peerId, "test", params);

		const data = await dataPromise;

		expect(data.sender.equals(remoteComponents.peerId)).toBe(true);
		expect(data.params).toStrictEqual(params);
	});

	it("calls the handler method on requests and returns the result", async () => {
		const params = new Uint8Array([1, 2, 3]);
		const response = new Uint8Array([2, 3, 4]);

		const dataPromise: Promise<{ params: Uint8Array, sender: PeerId }> = new Promise(resolve => {
			localRpc.addMethod("test", (params: Uint8Array, sender: PeerId) => {
				resolve({ params, sender });

				return response;
			});
		});

		const methodResponse = await remoteRpc.request(localComponents.peerId, "test", params);

		const data = await dataPromise;

		expect(data.sender.equals(remoteComponents.peerId)).toBe(true);
		expect(data.params).toStrictEqual(params);
		expect(methodResponse).toStrictEqual(response);
	});
});


src/RPCProtocol.ts
/* eslint-disable import/export */
/* eslint-disable complexity */
/* eslint-disable @typescript-eslint/no-namespace */
/* eslint-disable @typescript-eslint/no-unnecessary-boolean-literal-compare */

import { encodeMessage, decodeMessage, message } from 'protons-runtime'
import type { Uint8ArrayList } from 'uint8arraylist'
import type { Codec } from 'protons-runtime'

export interface RPCRequest {
  name: string
  params?: Uint8Array
  id?: number
}

export namespace RPCRequest {
  let _codec: Codec<RPCRequest>

  export const codec = (): Codec<RPCRequest> => {
    if (_codec == null) {
      _codec = message<RPCRequest>((obj, w, opts = {}) => {
        if (opts.lengthDelimited !== false) {
          w.fork()
        }

        if (opts.writeDefaults === true || obj.name !== '') {
          w.uint32(10)
          w.string(obj.name)
        }

        if (obj.params != null) {
          w.uint32(18)
          w.bytes(obj.params)
        }

        if (obj.id != null) {
          w.uint32(24)
          w.uint32(obj.id)
        }

        if (opts.lengthDelimited !== false) {
          w.ldelim()
        }
      }, (reader, length) => {
        const obj: any = {
          name: ''
        }

        const end = length == null ? reader.len : reader.pos + length

        while (reader.pos < end) {
          const tag = reader.uint32()

          switch (tag >>> 3) {
            case 1:
              obj.name = reader.string()
              break
            case 2:
              obj.params = reader.bytes()
              break
            case 3:
              obj.id = reader.uint32()
              break
            default:
              reader.skipType(tag & 7)
              break
          }
        }

        return obj
      })
    }

    return _codec
  }

  export const encode = (obj: RPCRequest): Uint8Array => {
    return encodeMessage(obj, RPCRequest.codec())
  }

  export const decode = (buf: Uint8Array | Uint8ArrayList): RPCRequest => {
    return decodeMessage(buf, RPCRequest.codec())
  }
}

export interface RPCError {
  code: number
  message: string
  data?: Uint8Array
}

export namespace RPCError {
  let _codec: Codec<RPCError>

  export const codec = (): Codec<RPCError> => {
    if (_codec == null) {
      _codec = message<RPCError>((obj, w, opts = {}) => {
        if (opts.lengthDelimited !== false) {
          w.fork()
        }

        if (opts.writeDefaults === true || obj.code !== 0) {
          w.uint32(8)
          w.int32(obj.code)
        }

        if (opts.writeDefaults === true || obj.message !== '') {
          w.uint32(18)
          w.string(obj.message)
        }

        if (obj.data != null) {
          w.uint32(26)
          w.bytes(obj.data)
        }

        if (opts.lengthDelimited !== false) {
          w.ldelim()
        }
      }, (reader, length) => {
        const obj: any = {
          code: 0,
          message: ''
        }

        const end = length == null ? reader.len : reader.pos + length

        while (reader.pos < end) {
          const tag = reader.uint32()

          switch (tag >>> 3) {
            case 1:
              obj.code = reader.int32()
              break
            case 2:
              obj.message = reader.string()
              break
            case 3:
              obj.data = reader.bytes()
              break
            default:
              reader.skipType(tag & 7)
              break
          }
        }

        return obj
      })
    }

    return _codec
  }

  export const encode = (obj: RPCError): Uint8Array => {
    return encodeMessage(obj, RPCError.codec())
  }

  export const decode = (buf: Uint8Array | Uint8ArrayList): RPCError => {
    return decodeMessage(buf, RPCError.codec())
  }
}

export interface RPCResponse {
  id: number
  result?: Uint8Array
  error?: RPCError
}

export namespace RPCResponse {
  let _codec: Codec<RPCResponse>

  export const codec = (): Codec<RPCResponse> => {
    if (_codec == null) {
      _codec = message<RPCResponse>((obj, w, opts = {}) => {
        if (opts.lengthDelimited !== false) {
          w.fork()
        }

        if (opts.writeDefaults === true || obj.id !== 0) {
          w.uint32(8)
          w.uint32(obj.id)
        }

        if (obj.result != null) {
          w.uint32(18)
          w.bytes(obj.result)
        }

        if (obj.error != null) {
          w.uint32(26)
          RPCError.codec().encode(obj.error, w, {
            writeDefaults: false
          })
        }

        if (opts.lengthDelimited !== false) {
          w.ldelim()
        }
      }, (reader, length) => {
        const obj: any = {
          id: 0
        }

        const end = length == null ? reader.len : reader.pos + length

        while (reader.pos < end) {
          const tag = reader.uint32()

          switch (tag >>> 3) {
            case 1:
              obj.id = reader.uint32()
              break
            case 2:
              obj.result = reader.bytes()
              break
            case 3:
              obj.error = RPCError.codec().decode(reader, reader.uint32())
              break
            default:
              reader.skipType(tag & 7)
              break
          }
        }

        return obj
      })
    }

    return _codec
  }

  export const encode = (obj: RPCResponse): Uint8Array => {
    return encodeMessage(obj, RPCResponse.codec())
  }

  export const decode = (buf: Uint8Array | Uint8ArrayList): RPCResponse => {
    return decodeMessage(buf, RPCResponse.codec())
  }
}

export interface RPCMessage {
  request?: RPCRequest
  response?: RPCResponse
}

export namespace RPCMessage {
  let _codec: Codec<RPCMessage>

  export const codec = (): Codec<RPCMessage> => {
    if (_codec == null) {
      _codec = message<RPCMessage>((obj, w, opts = {}) => {
        if (opts.lengthDelimited !== false) {
          w.fork()
        }

        if (obj.request != null) {
          w.uint32(10)
          RPCRequest.codec().encode(obj.request, w, {
            writeDefaults: false
          })
        }

        if (obj.response != null) {
          w.uint32(18)
          RPCResponse.codec().encode(obj.response, w, {
            writeDefaults: false
          })
        }

        if (opts.lengthDelimited !== false) {
          w.ldelim()
        }
      }, (reader, length) => {
        const obj: any = {}

        const end = length == null ? reader.len : reader.pos + length

        while (reader.pos < end) {
          const tag = reader.uint32()

          switch (tag >>> 3) {
            case 1:
              obj.request = RPCRequest.codec().decode(reader, reader.uint32())
              break
            case 2:
              obj.response = RPCResponse.codec().decode(reader, reader.uint32())
              break
            default:
              reader.skipType(tag & 7)
              break
          }
        }

        return obj
      })
    }

    return _codec
  }

  export const encode = (obj: RPCMessage): Uint8Array => {
    return encodeMessage(obj, RPCMessage.codec())
  }

  export const decode = (buf: Uint8Array | Uint8ArrayList): RPCMessage => {
    return decodeMessage(buf, RPCMessage.codec())
  }
}


src/MessageHandler.ts
import type { ConnectionManager } from "@libp2p/interface-connection-manager";
import type { Connection, Stream } from "@libp2p/interface-connection";
import type { Registrar } from "@libp2p/interface-registrar";
import type { PeerId } from "@libp2p/interface-peer-id";
import * as lp from "it-length-prefixed";
import { pipe } from "it-pipe";
import { pushable, Pushable } from "it-pushable";

export interface MessageHandlerComponents {
	connectionManager: ConnectionManager
	registrar: Registrar
}

export interface MessageHandlerOpts {
	protocol: string
}

type Handler = (message: Uint8Array, peer: PeerId) => void;

export class MessageHandler {
	private readonly components: MessageHandlerComponents;
	private readonly options: MessageHandlerOpts;
	private readonly writers = new Map<string, Pushable<Uint8Array>>();
	private readonly handlers = new Set<Handler>();

	constructor (components: MessageHandlerComponents, options: Partial<MessageHandlerOpts> = {}) {
		this.components = components;

		this.options = {
			protocol: options.protocol ?? "/message-handler/0.0.1",
		};
	}

	async start () {
		await this.components.registrar.handle(this.options.protocol, async ({ stream, connection }) => {
			this.handleStream(stream, connection);
		});
	}

	async stop () {
		await this.components.registrar.unhandle(this.options.protocol);
	}

	async send (peer: PeerId, message: Uint8Array) {
		const writer = await this.establishStream(peer);

		writer.push(message);
	}

	handle (handler: Handler) {
		this.handlers.add(handler);
	}

	unhandle (handler: Handler) {
		this.handlers.delete(handler);
	}

	// Establish a stream to a peer reusing an existing one if it already exists.
	private async establishStream (peer: PeerId) {
		const connection = this.components.connectionManager.getConnections().find(c => c.remotePeer.equals(peer));

		if (connection == null) {
			throw new Error("not connected");
		}

		for (const stream of connection.streams) {
			if (this.writers.has(stream.id)) {
				// We already have a stream open.
				return this.writers.get(stream.id)!;
			}
		}

		const stream = await connection.newStream(this.options.protocol);

		return this.handleStream(stream, connection);
	}

	private handleStream (stream: Stream, connection: Connection) {
		const handlers = this.handlers;
		const peerId = connection.remotePeer.toString();

		// Handle inputs.
		pipe(stream, lp.decode(), async function (source) {
			for await (const message of source) {
				for (const handler of handlers) {
					handler(message.subarray(), connection.remotePeer);
				}
			}
		}).catch(() => {
			// Do nothing
		});

		// Don't pipe events through the same connection
		if (this.writers.has(peerId)) {
			return this.writers.get(peerId)!;
		}

		const writer = pushable();

		this.writers.set(peerId, writer);

		// Handle outputs.
		(async () => {
			try {
				await pipe(writer, lp.encode(), stream);
			} finally {
				this.writers.delete(peerId);
			}
		})();

		return writer;
	}
}

export const createMessageHandler = (options?: Partial<MessageHandlerOpts>) => (components: MessageHandlerComponents) => new MessageHandler(components, options);


src/RPC.ts
import type { PeerId } from "@libp2p/interface-peer-id";
import type { Startable } from "@libp2p/interfaces/startable";
import { logger } from "@libp2p/logger";
import {
	createMessageHandler,
	MessageHandler,
	MessageHandlerComponents
} from "@organicdesign/libp2p-message-handler";
import { RPCMessage } from "./RPCProtocol.js";
import * as Messages from "./RPCMessages.js";
import { RPCException } from "./RPCException.js";

const log = {
	general: logger("libp2p:rpc")
};

export interface RPCOpts {
	protocol: string,
	timeout: number
}

export interface RPCComponents extends MessageHandlerComponents {}

export type RPCMethod = (params: Uint8Array | undefined, sender: PeerId) => Promise<Uint8Array | void> | Uint8Array | void;

interface Resolver {
	resolve: (result?: Uint8Array) => void
	reject: (error: RPCException) => void
}

export class RPC implements Startable {
	private readonly options: RPCOpts;
	private readonly methods = new Map<string, RPCMethod>();
	private readonly msgPromises = new Map<number, Resolver>();
	private readonly handler: MessageHandler;
	private started = false;

	private readonly genMsgId = (() => {
		let id = 0;

		return () => id++;
	})();

	constructor (components: RPCComponents, options: Partial<RPCOpts> = {}) {
		this.options = {
			protocol: options.protocol ?? "/libp2p-rpc/0.0.1",
			timeout: options.timeout ?? 5000
		};

		this.handler = createMessageHandler({ protocol: this.options.protocol })(components);
	}

	async start (): Promise<void> {
		if (this.isStarted()) {
			return;
		}

		await this.handler.start();

		this.handler.handle((message, peer) => {
			this.handleMessage(RPCMessage.decode(message), peer);
		});

		this.started = true;

		log.general("started");
	}

	async stop (): Promise<void> {
		if (!this.isStarted()) {
			return;
		}

		await this.handler.stop();

		// Reject the open promises.
		for (const promise of this.msgPromises.values()) {
			promise.reject(new RPCException("RPC module stopped", -32001));
		}

		this.msgPromises.clear();
		this.methods.clear();

		this.started = false;

		log.general("stopped");
	}

	isStarted(): boolean {
		return this.started;
	}

	addMethod (name: string, method: RPCMethod): void {
		this.methods.set(name, method);
	}

	hasMethod (name: string): boolean {
		return this.methods.has(name);
	}

	removeMethod (name: string): boolean {
		return this.methods.delete(name);
	}

	async request (peer: PeerId, name: string, params?: Uint8Array): Promise<Uint8Array | undefined> {
		const messageId = this.genMsgId();

		try {
			await this.handler.send(Messages.createRequest(name, messageId, params), peer);
		} catch (error) {
			log.general.error("failed to send message: %o", error);

			const newError = new RPCException(error.message, -32000);

			throw newError;
		}

		log.general("request '%s' on peer: %p", name, peer);

		return await new Promise((resolve, reject) => {
			if (this.options.timeout < 0) {
				return;
			}

			const timeoutError = new RPCException("Request timed out", -32003);
			const timeout = setTimeout(() => reject(timeoutError), this.options.timeout);

			this.msgPromises.set(messageId, {
				resolve (result) {
					clearTimeout(timeout);
					resolve(result);
				},
				reject (error) {
					clearTimeout(timeout);
					reject(error);
				}
			});
		});
	}

	notify (peer: PeerId, name: string, params?: Uint8Array): void {
		this.handler.send(Messages.createNotification(name, params), peer).catch(() => {});

		log.general("notify '%s' on peer: %p", name, peer);
	}

	// Handle receiving a messsage calling RPC methods or resolving responses.
	private async handleMessage (message: RPCMessage, peer: PeerId) {
		const { request, response } = message;

		if (request != null) {
			const method = this.methods.get(request.name);

			if (!method) {
				if (request.id == null) {
					return;
				}

				return await this.handler.send(Messages.createMethodNotFoundError(request.id), peer);
			}

			let result: Uint8Array | undefined;
			let error: RPCException | null = null;

			try {
				log.general("method '%s' called by peer: %p", request.name, peer);
				result = await method(request.params, peer) ?? undefined;
			} catch (err) {
				log.general.error("method '%s' threw error: %o", err);

				if (err instanceof RPCException) {
					error = err;
				} else if (err instanceof Error) {
					error = new RPCException(err.message, 0);
					error.stack = err.stack;
				} else {
					try {
						error = new RPCException(JSON.stringify(err), -32002);
					} catch (err) {
						error = new RPCException("Unknown error", -32002);
					}
				}
			}

			if (request.id == null) {
				return;
			}

			if (error != null) {
				return await this.handler.send(Messages.createError(request.id, error.message, error.code), peer);
			}

			return await this.handler.send(Messages.createResponse(request.id, result), peer);
		}

		if (response) {
			const resolver = this.msgPromises.get(response.id);

			if (resolver == null) {
				return;
			}

			this.msgPromises.delete(response.id);

			if (response.error == null) {
				return resolver.resolve(response.result);
			}

			resolver.reject(new RPCException(response.error.message, response.error.code, response.error.data));
		}
	}
}

export const createRPC = (options?: Partial<RPCOpts>) => (components: RPCComponents) => new RPC(components, options);
